Improve complex support in common_nn test machinery

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50594 Padding: support complex dtypes
* **#50593 Improve complex support in common_nn test machinery**

There are no equivalent to torch.FloatTensor, torch.cuda.FloatTensor for complex
types. So `get_gpu_type` and `get_cpu_type` are broken for complex tensors.

Also found a few places that explicitly cast inputs to floating point types,
which would drop the imaginary component before running the test.

Differential Revision: [D25954050](https://our.internmc.facebook.com/intern/diff/D25954050)

peterbell10 committed:
Improve complex support in common_nn test machinery

There are no equivalent to torch.FloatTensor, torch.cuda.FloatTensor for complex
types. So `get_gpu_type` and `get_cpu_type` are broken for complex tensors.

Also found a few places that explicitly cast inputs to floating point types,
which would drop the imaginary component before running the test.

[ghstack-poisoned]

peterbell10 committed:
Update on "Improve complex support in common_nn test machinery"


There are no equivalent to torch.FloatTensor, torch.cuda.FloatTensor for complex
types. So `get_gpu_type` and `get_cpu_type` are broken for complex tensors.

Also found a few places that explicitly cast inputs to floating point types,
which would drop the imaginary component before running the test.

[ghstack-poisoned]

anjali411 commented:
we had a similar [discussion](https://github.com/pytorch/pytorch/pull/49912#discussion_r551612767) in this PR adding complex support for L1 loss function https://github.com/pytorch/pytorch/pull/49912

anjali411 commented on torch/testing/_internal/common_utils.py:
```diff
@@ -597,29 +597,19 @@ def wrapper(*args, **kwargs):
-        with torch.no_grad():
-            res = obj.clone().type(t)
-            res.requires_grad = obj.requires_grad
+        res = obj.cuda()
```
to make the change minimal, can we just do:
```
t = type_map.get(obj.dtype, obj.dtype)
with torch.no_grad():
        res = obj.clone().to(dtype=t, device="cuda")
        res.requires_grad = obj.requires_grad
return res
```

anjali411 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -5120,15 +5121,31 @@ def assert_module_parameters_are(tensor_type, device_id=None):
+                else:
+                    return tensor
+
+            def to_half(x):
```
I don't think we should add `to_half` for complex tensors, since the support for `torch.complex32` is bare minimum right now. In fact, @mruberry and I both agreed we should probably disable `torch.complex32` for 1.8 release.

peterbell10 committed:
Update on "Improve complex support in common_nn test machinery"


There are no equivalent to torch.FloatTensor, torch.cuda.FloatTensor for complex
types. So `get_gpu_type` and `get_cpu_type` are broken for complex tensors.

Also found a few places that explicitly cast inputs to floating point types,
which would drop the imaginary component before running the test.

[ghstack-poisoned]

peterbell10 commented:
Rebased on #49912 after it was merged and removed the reference to `torch.complex32`.

peterbell10 committed:
Update on "Improve complex support in common_nn test machinery"


There are no equivalent to torch.FloatTensor, torch.cuda.FloatTensor for complex
types. So `get_gpu_type` and `get_cpu_type` are broken for complex tensors.

Also found a few places that explicitly cast inputs to floating point types,
which would drop the imaginary component before running the test.

[ghstack-poisoned]

peterbell10 committed:
Update on "Improve complex support in common_nn test machinery"


There are no equivalent to torch.FloatTensor, torch.cuda.FloatTensor for complex
types. So `get_gpu_type` and `get_cpu_type` are broken for complex tensors.

Also found a few places that explicitly cast inputs to floating point types,
which would drop the imaginary component before running the test.

[ghstack-poisoned]

mruberry reviewed (approved):
Looks OK to me. Let's let @anjali411 take a look, too.

anjali411 reviewed (approved):
LGTM!

peterbell10 committed:
Update on "Improve complex support in common_nn test machinery"


There are no equivalent to torch.FloatTensor, torch.cuda.FloatTensor for complex
types. So `get_gpu_type` and `get_cpu_type` are broken for complex tensors.

Also found a few places that explicitly cast inputs to floating point types,
which would drop the imaginary component before running the test.

Differential Revision: [D25954050](https://our.internmc.facebook.com/intern/diff/D25954050)

[ghstack-poisoned]

anjali411 committed:
Update on "Improve complex support in common_nn test machinery"


There are no equivalent to torch.FloatTensor, torch.cuda.FloatTensor for complex
types. So `get_gpu_type` and `get_cpu_type` are broken for complex tensors.

Also found a few places that explicitly cast inputs to floating point types,
which would drop the imaginary component before running the test.

Differential Revision: [D25954050](https://our.internmc.facebook.com/intern/diff/D25954050)

[ghstack-poisoned]

peterbell10 committed:
Update on "Improve complex support in common_nn test machinery"


There are no equivalent to torch.FloatTensor, torch.cuda.FloatTensor for complex
types. So `get_gpu_type` and `get_cpu_type` are broken for complex tensors.

Also found a few places that explicitly cast inputs to floating point types,
which would drop the imaginary component before running the test.

Differential Revision: [D25954050](https://our.internmc.facebook.com/intern/diff/D25954050)

[ghstack-poisoned]

peterbell10 commented on torch/testing/_internal/jit_metaprogramming_utils.py:
```diff
@@ -543,7 +549,7 @@ def try_get_nn_module_compiled_mod_and_inputs(*args, **kwargs):
         input = input + (kwargs['target_fn'](),)
 
-    args_variable, kwargs_variable = create_input(input)
+    args_variable, kwargs_variable = create_input(input, dtype=input_dtype)
```
@anjali411 this should fix the test failure in #50594.
